java/src/main/java/ai/onnxruntime/OnnxRuntime.java

```diff
@@ -5,175 +5,91 @@
+  private static void load(String library) throws IOException {
+    try {
+      logger.log(Level.FINE, "Attempting to load native library '" + library + "' from lib path");
+      System.loadLibrary(library);
+      logger.log(Level.FINE, "Loaded native library '" + library + "' from lib path");
+      return;
+    } catch (Exception | Error e) {
+      logger.log(Level.FINE, "Native library '" + library + "' failed to load from lib path", e);
+    }
+    // generate a platform specific library name
+    // replace Mac's jnilib extension to dylib
+    String libraryFileName = System.mapLibraryName(library).replace("jnilib", "dylib");
+    String resourcePath = "/ai/onnxruntime/native/" + libraryFileName;
+    logger.log(
+        Level.FINE,
+        "Attempting to load native library '" + library + "' from resource path " + resourcePath);
+    InputStream is = OnnxRuntime.class.getResourceAsStream(resourcePath);
+    if (is == null) {
+      logger.log(Level.FINE, "Native library '" + library + "' was not found in resources");
+      return;
+    }
 
-            if (filename != null && prefix.length() >= 3) {
-                File temp = File.createTempFile(prefix, suffix);
-                if (debugLogging) {
-                    logger.info("Writing " + path + " out to " + temp.getAbsolutePath());
-                }
-                temp.deleteOnExit();
-                if (!temp.exists()) {
-                    throw new FileNotFoundException("File " + temp.getAbsolutePath() + " does not exist.");
-                } else {
-                    byte[] buffer = new byte[1024];
-                    try (InputStream is = OnnxRuntime.class.getResourceAsStream(path)) {
-                        if (is == null) {
-                            throw new FileNotFoundException("File " + path + " was not found inside JAR.");
-                        } else {
-                            int readBytes;
-                            try (FileOutputStream os = new FileOutputStream(temp)) {
-                                while ((readBytes = is.read(buffer)) != -1) {
-                                    os.write(buffer, 0, readBytes);
-                                }
-                            }
-                            return temp.getAbsolutePath();
-                        }
-                    }
-                }
-            } else {
-                throw new IllegalArgumentException("The filename has to be at least 3 characters long.");
-            }
+    File temp = File.createTempFile("javaload_", libraryFileName);
+    try {
+      byte[] buffer = new byte[1024];
+      int readBytes;
+      try (FileOutputStream os = new FileOutputStream(temp)) {
+        while ((readBytes = is.read(buffer)) != -1) {
+          os.write(buffer, 0, readBytes);
         }
+      }
+      System.load(temp.getAbsolutePath());
+      logger.log(Level.FINE, "Loaded native library '" + library + "' from resources");
+    } finally {
+      temp.delete();
```

Comment:
This probably should be `temp.deleteOnExit()`, I'm not sure that the delete trigger will take effect if the library has been loaded.

Reply:
If the temp is writable by the other users, it's kind of a security hole.

For example, it could be:
1. A privileged  user got the onnxruntime.jar and started to run it.
2. A temp file was created under the temp directory
3. A hacker found the file and replaced it with something else
4. The dll(or so) was loaded by the privileged user and run.
5. The hacker gain unauthorized access.

Reply:
The internet is pretty undecided about the proper way to load shared libraries from the classpath. I've worked with Netty's native library loader in the past and they seem to employ the very same tempfile method. I don't know if that is best practice, but that's been around for a while. They also by default do a delete like I did. There is a deleteOnExit option, but I do not know why that use case exists. Do you have a deleteOnExit use case? Once the System.load call is done, the library should be in virtual memory and the file can be removed fully. https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L234

In regards to security in the scenario given, I think timing (the short amount of time the file exists) and the random number in the temp filename that Java adds may shield us for the most part. I suppose a hacker could try to predict the random number and somehow maybe listen to file changes and just perfectly swap the file right at just the right moment. a quick internet search of `java file.createTempFile security` did not turn up much except for advising to not create directories.

Reply:
As to the security issue, I think by default the file on Linux and macOS should only be writable by the user who created it, so I'm not too worried about that. If there is a privileged attacker on the system who can overwrite users' files then there are bigger problems. A quick check with jshell on Windows should determine what default permissions are applied to the temp file.

You can control the permissions that are used to create the file via [this method](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createTempFile-java.nio.file.Path-java.lang.String-java.lang.String-java.nio.file.attribute.FileAttribute...-), but I'm not sure if the allowable attributes differ across Windows, macOS and Linux.

I think some OSs don't allow you to delete a file that has been dlopened until it's no longer in use. Windows has this issue, it won't let you delete a dll that's in use until the process has gone away (though actually on Windows that means even the `deleteOnExit` hook isn't enough).

Reply:
I do not have a windows runtime available to me to test this out, but I took the netty loader as inspiration and will deleteOnExit if the delete fails (presumably because the file is still open). It should fail in the CI job I believe if this is not the right solution.

Reply:
I don't think it will fail in the CI, as the deleteOnExit call isn't going to fail then, and when the shutdown hook happens the JVM is going away anyway so I think it gets swallowed. But this isn't too much of an issue as deleteOnExit is the best you can do in the JDK.